### ISSUE TITLE
Phase 2: Feature Engineering and Baseline Model Implementation

### ISSUE THREAD

**Issue by dhar174** [dhar174/steel_defect_demo#5]
## **Note to Assistant: if the sub-issues are all completed, simply check over each individual sub-issue thoroughly for blocking/breaking errors, check over all tasks in the parent issue, and just overall perform final polish, ensuring compatibility with all other parts of the app (where interactions between components exists). We need a holistic, finished product. Also note: No large/major refactors if it isnt to fix app-breaking issues!**

# Phase 2: Feature Engineering and Baseline Model Implementation

## Overview
Implement comprehensive feature engineering pipeline and develop the baseline XGBoost model for steel casting defect prediction. This phase builds upon the synthetic data generated in Phase 1 to create a robust machine learning foundation.

## Objectives
- Develop a comprehensive feature engineering pipeline that extracts meaningful patterns from time series data
- Implement and train a baseline XGBoost model for defect prediction
- Establish model evaluation framework with proper metrics and validation
- Create feature analysis workflows to understand data patterns

### 2.1 Feature Engineering Pipeline Implementation
- **File**: `src/features/feature_engineer.py`
- **Dependencies**: pandas, numpy, scikit-learn
- **Key Features**:
  - Statistical features (mean, std, min, max, median, percentiles) for each sensor
  - Stability features (spike counts, excursion frequencies, variation coefficients)
  - Duration features (time spent at extremes, threshold crossings)
  - Interaction features (cross-sensor correlations, ratios, differences)
  - Temporal features (trends, gradients, change rates)

### 2.2 Baseline XGBoost Model Implementation
- **File**: `src/models/baseline_model.py`
- **Dependencies**: xgboost, scikit-learn, joblib
- **Key Features**:
  - Configurable XGBoost classifier with hyperparameter management
  - Cross-validation with stratified K-fold
  - Feature importance analysis and selection
  - Model serialization and persistence
  - Comprehensive logging and metrics tracking

### 2.3 Model Training Pipeline
- **File**: `src/models/model_trainer.py`
- **Dependencies**: scikit-learn, yaml
- **Functionality**:
  - Automated training pipeline with configuration management
  - Data preprocessing and feature scaling
  - Hyperparameter optimization with grid search
  - Early stopping and overfitting prevention
  - Training history and performance tracking

### 2.4 Model Evaluation Framework
- **File**: `src/models/model_evaluator.py`
- **Dependencies**: scikit-learn, matplotlib, seaborn
- **Features**:
  - Comprehensive metric calculation (AUC-ROC, AUC-PR, F1, Precision, Recall)
  - Confusion matrix and classification report generation
  - ROC and Precision-Recall curve plotting
  - Feature importance visualization
  - Cross-validation performance analysis

### 2.5 Training Script
- **File**: `scripts/train_baseline_model.py`
- **Functionality**:
  - Command-line interface for model training
  - Configuration loading and validation
  - Data loading and preprocessing coordination
  - Model training execution with progress tracking
  - Results saving and model artifacts management

### 2.6 Feature Analysis Notebook
- **File**: `notebooks/02_feature_analysis.ipynb`
- **Analysis Components**:
  - Feature distribution analysis for normal vs defect casts
  - Feature correlation matrix and heatmaps
  - Feature importance ranking and interpretation
  - Univariate and bivariate feature analysis
  - Feature selection recommendations

### Feature Engineering Requirements
- **Statistical Features**: 7 statistics × 5 sensors = 35 features
- **Stability Features**: 4 stability metrics × 5 sensors = 20 features
- **Duration Features**: 3 duration metrics × 5 sensors = 15 features
- **Interaction Features**: 10 cross-sensor combinations = 10 features
- **Total Feature Count**: ~80 engineered features

### Model Configuration
```yaml
baseline_model:
  algorithm: "xgboost"
  parameters:
    n_estimators: 100
    max_depth: 6
    learning_rate: 0.1
    subsample: 0.8
    colsample_bytree: 0.8
    random_state: 42
  
  validation:
    cv_folds: 5
    early_stopping_rounds: 10
    eval_metric: "auc"
```

### Performance Targets
- **AUC-ROC**: > 0.85 on test set
- **AUC-PR**: > 0.70 on test set
- **Training Time**: < 5 minutes on standard hardware
- **Inference Time**: < 100ms per prediction
- **Memory Usage**: < 2GB during training

## Deliverables
1. Complete `CastingFeatureEngineer` class with all feature extraction methods
2. Functional `BaselineXGBoostModel` class with training and prediction capabilities
3. Comprehensive model evaluation framework with metrics and visualizations
4. Feature analysis notebook with detailed insights
5. Training script with command-line interface
6. Unit tests for feature engineering and model components (`tests/test_feature_engineering.py`, `tests/test_models.py`)
7. Model artifacts and performance reports

## Testing Requirements
- Unit tests for feature engineering methods with edge case coverage
- Model training and prediction functionality tests
- Feature engineering pipeline integration tests
- Performance benchmarking tests
- Data validation and quality checks

## Dependencies
```
pandas>=1.5.0
numpy>=1.23.0
scikit-learn>=1.1.0
xgboost>=1.7.0
joblib>=1.2.0
matplotlib>=3.6.0
seaborn>=0.12.0
pyyaml>=6.0
jupyter>=1.0.0
pytest>=7.2.0
```

## Execution Commands
```bash
# Train baseline model
python scripts/train_baseline_model.py

# Run feature analysis
jupyter notebook notebooks/02_feature_analysis.ipynb

# Run tests
python -m pytest tests/test_feature_engineering.py -v
python -m pytest tests/test_models.py -v

# Evaluate model performance
python scripts/evaluate_models.py --model baseline
```

## Success Criteria
- [ ] Feature engineering pipeline extracts 80+ meaningful features from time series data
- [ ] Baseline XGBoost model achieves AUC-ROC > 0.85 on test set
- [ ] Feature importance analysis identifies top predictive features
- [ ] Cross-validation shows consistent performance across folds
- [ ] All unit tests pass with >80% code coverage
- [ ] Training script executes successfully with proper logging
- [ ] Feature analysis notebook provides actionable insights
- [ ] Model artifacts are properly saved and can be loaded for inference

## Prerequisites
- [ ] Phase 1 synthetic data generation completed
- [ ] Dataset contains 1,200+ casting sequences with proper labels
- [ ] Project structure and configuration files in place
- [ ] Development environment set up with required dependencies
